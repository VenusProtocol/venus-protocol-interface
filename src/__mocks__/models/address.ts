const address = '0x3d759121234cd36F8124C21aFe1c6852d2bEd848';
export default address;
