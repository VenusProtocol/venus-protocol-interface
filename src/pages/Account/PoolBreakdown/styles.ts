import { css } from '@emotion/react';
import { useTheme } from '@mui/material';

export const useStyles = () => {
  const theme = useTheme();

  return {
    title: css`
      display: flex;
      align-items: center;
      margin-bottom: ${theme.spacing(4)};
    `,
    marketName: css`
      margin-right: ${theme.spacing(3)};
    `,
    statsContainer: css`
      display: flex;
      justify-content: space-between;
      margin-bottom: ${theme.spacing(6)};

      ${theme.breakpoints.down('xl')} {
        display: block;
        padding: 0;
        background-color: transparent;
      }

      ${theme.breakpoints.down('md')} {
        margin-bottom: ${theme.spacing(10)};
      }
    `,
    getNetApyColor: ({ netApyPercentage }: { netApyPercentage: number }) =>
      netApyPercentage > 0 ? theme.palette.interactive.success : theme.palette.interactive.error,
    cellGroup: css`
      padding: 0;

      ${theme.breakpoints.down('xl')} {
        margin-bottom: ${theme.spacing(2)};
      }
    `,
    accountHealth: css`
      ${theme.breakpoints.down('xl')} {
        padding: ${theme.spacing(4)};
        background-color: ${theme.palette.background.paper};
        border-radius: ${theme.shape.borderRadius.large}px;
      }
    `,
    accountHealthProgressBar: css`
      margin-bottom: ${theme.spacing(3)};
    `,
    accountHealthFooter: css`
      display: flex;
      justify-content: flex-end;
      align-items: center;
    `,
    tooltip: css`
      display: flex;
    `,
    infoIcon: css`
      cursor: help;
    `,
    inlineLabel: css`
      margin-right: ${theme.spacing(1)};
    `,
    progressBar: css`
      margin-bottom: ${theme.spacing(3)};
    `,
    shieldIcon: css`
      margin-right: ${theme.spacing(2)};
    `,
    safeLimit: css`
      margin-right: ${theme.spacing(2)};
    `,
  };
};
