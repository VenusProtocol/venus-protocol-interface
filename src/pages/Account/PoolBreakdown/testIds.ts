const TEST_IDS = {
  tables: 'account-pool-breakdown-tables',
};

export default TEST_IDS;
