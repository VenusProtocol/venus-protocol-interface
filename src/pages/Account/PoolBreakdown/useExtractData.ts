import BigNumber from 'bignumber.js';
import { useMemo } from 'react';
import { Asset } from 'types';
import {
  calculateCollateralValue,
  calculateDailyEarningsCents,
  calculateNetApy,
  calculateYearlyEarningsForAssets,
  convertTokensToWei,
  formatCentsToReadableValue,
  formatToReadablePercentage,
} from 'utilities';

import { SAFE_BORROW_LIMIT_PERCENTAGE } from 'constants/safeBorrowLimitPercentage';

const useExtractData = ({ assets, includeXvs }: { assets: Asset[]; includeXvs: boolean }) =>
  useMemo(() => {
    const { totalBorrowCents, totalSupplyCents, borrowLimitCents } = assets.reduce(
      (acc, asset) => ({
        totalBorrowCents: acc.totalBorrowCents.plus(
          asset.userBorrowBalanceTokens.times(asset.tokenPriceDollars).times(100),
        ),
        totalSupplyCents: acc.totalSupplyCents.plus(
          asset.userSupplyBalanceTokens.times(asset.tokenPriceDollars).times(100),
        ),
        borrowLimitCents: asset.isCollateralOfUser
          ? acc.borrowLimitCents.plus(
              calculateCollateralValue({
                amountWei: convertTokensToWei({
                  value: asset.userSupplyBalanceTokens,
                  token: asset.vToken.underlyingToken,
                }),
                token: asset.vToken.underlyingToken,
                tokenPriceDollars: asset.tokenPriceDollars,
                collateralFactor: asset.collateralFactor,
              }).times(100),
            )
          : acc.borrowLimitCents,
      }),
      {
        totalSupplyCents: new BigNumber(0),
        totalBorrowCents: new BigNumber(0),
        borrowLimitCents: new BigNumber(0),
      },
    );

    const yearlyEarningsCents = calculateYearlyEarningsForAssets({
      assets,
      includeXvs,
    });

    const dailyEarningsCentsTmp =
      yearlyEarningsCents && calculateDailyEarningsCents(yearlyEarningsCents).toNumber();

    const netApyPercentageTmp =
      yearlyEarningsCents &&
      calculateNetApy({
        supplyBalanceCents: totalSupplyCents,
        yearlyEarningsCents,
      });

    const safeBorrowLimitCentsTmp = borrowLimitCents.multipliedBy(
      SAFE_BORROW_LIMIT_PERCENTAGE / 100,
    );

    const readableSafeBorrowLimitTmp = formatCentsToReadableValue({
      value: safeBorrowLimitCentsTmp,
    });

    const safeBorrowLimitPercentageTmp = formatToReadablePercentage(
      safeBorrowLimitCentsTmp.multipliedBy(100).dividedBy(borrowLimitCents),
    );

    return {
      dailyEarningsCents: dailyEarningsCentsTmp,
      netApyPercentage: netApyPercentageTmp,
      readableSafeBorrowLimit: readableSafeBorrowLimitTmp,
      safeBorrowLimitPercentage: safeBorrowLimitPercentageTmp,
      totalBorrowCents,
      totalSupplyCents,
      borrowLimitCents,
    };
  }, [JSON.stringify(assets), includeXvs]);

export default useExtractData;
