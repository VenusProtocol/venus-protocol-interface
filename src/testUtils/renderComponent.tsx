import React from 'react';
import { render } from '@testing-library/react';
import { QueryClient, QueryClientProvider } from 'react-query';
import { BrowserRouter, Route, Switch } from 'react-router-dom';
import { ToastContainer } from 'react-toastify';

import { Web3Wrapper } from 'clients/web3';
import { AuthContext, IAuthContextValue } from 'context/AuthContext';
import { SuccessfulTransactionModalProvider } from 'context/SuccessfulTransactionModalContext';
import Theme from 'theme';
import { RefreshContextProvider } from 'context/RefreshContext';
import { VaiContext, IVaiContextValue } from 'context/VaiContext';
import { MuiThemeProvider } from 'theme/MuiThemeProvider/MuiThemeProvider';
import BigNumber from 'bignumber.js';

const renderComponent = (
  children: React.ReactElement | (() => React.ReactElement),
  {
    authContextValue = {},
    vaiContextValue = {},
  }: {
    authContextValue?: Partial<IAuthContextValue>;
    vaiContextValue?: Partial<IVaiContextValue>;
  } = {},
) => {
  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
        cacheTime: 0,
      },
    },
  });
  const defaultAuthContextValues = {
    login: jest.fn(),
    logOut: jest.fn(),
    openAuthModal: jest.fn(),
    closeAuthModal: jest.fn(),
    account: undefined,
    ...authContextValue,
  };
  const defaultVaiContextValues = {
    userVaiMinted: new BigNumber(0),
    userVaiBalance: new BigNumber(0),
    userVaiEnabled: false,
    mintableVai: new BigNumber(0),
    ...vaiContextValue,
  };

  const renderRes = render(
    <Theme>
      <Web3Wrapper>
        <QueryClientProvider client={queryClient}>
          <MuiThemeProvider>
            <AuthContext.Provider value={defaultAuthContextValues}>
              <RefreshContextProvider>
                <VaiContext.Provider value={defaultVaiContextValues}>
                  <SuccessfulTransactionModalProvider>
                    <BrowserRouter>
                      <ToastContainer />

                      <Switch>
                        <Route
                          path="/"
                          component={typeof children === 'function' ? children : () => children}
                        />
                      </Switch>
                    </BrowserRouter>
                  </SuccessfulTransactionModalProvider>
                </VaiContext.Provider>
              </RefreshContextProvider>
            </AuthContext.Provider>
          </MuiThemeProvider>
        </QueryClientProvider>
      </Web3Wrapper>
    </Theme>,
  );

  return {
    ...renderRes,
    queryClient,
  };
};

export default renderComponent;
