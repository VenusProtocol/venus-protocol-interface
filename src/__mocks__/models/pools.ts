import { Pool } from 'types';

import { assetData } from '__mocks__/models/asset';

export const poolData: Pool[] = [
  {
    id: '1',
    assets: assetData,
    name: 'Venus',
    riskLevel: 'MINIMAL',
    description: 'Fake description 1',
    isIsolated: false,
  },
  {
    id: '2',
    assets: assetData,
    name: 'Metaverse',
    riskLevel: 'VERY_HIGH',
    description: 'Fake description 2',
    isIsolated: true,
  },
  {
    id: '3',
    assets: assetData,
    name: 'Gaming',
    riskLevel: 'MEDIUM',
    description: 'Fake description 3',
    isIsolated: true,
  },
];
