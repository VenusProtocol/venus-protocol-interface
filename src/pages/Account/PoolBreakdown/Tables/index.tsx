/** @jsxImportSource @emotion/react */
import { Typography } from '@mui/material';
import Paper from '@mui/material/Paper';
import { ButtonGroup } from 'components';
import React, { useMemo, useState } from 'react';
import { useTranslation } from 'translation';
import { Asset } from 'types';

import { MarketTable, MarketTableProps } from 'containers/MarketTable';
import { useHideMdDownCss, useHideXlDownCss, useShowXlDownCss } from 'hooks/responsive';

import TEST_IDS from '../testIds';
import { useStyles } from './styles';

export interface TablesProps {
  assets: Asset[];
}

export const Tables: React.FC<TablesProps> = ({ assets }) => {
  const styles = useStyles();
  const { t } = useTranslation();
  const [activeTabIndex, setActiveTabIndex] = useState(0);

  const hideXlDownCss = useHideXlDownCss();
  const showXlDownCss = useShowXlDownCss();
  const hideMdDownCss = useHideMdDownCss();

  const marketTableProps: {
    supply: MarketTableProps;
    borrow: MarketTableProps;
  } = useMemo(
    () => ({
      supply: {
        assets: assets.filter(asset => asset.userSupplyBalanceTokens.isGreaterThan(0)),
        marketType: 'supply',
        breakpoint: 'md',
        columns: ['asset', 'supplyApyLtv', 'userSupplyBalance', 'collateral'],
        initialOrder: {
          orderBy: 'userSupplyBalance',
          orderDirection: 'desc',
        },
      },
      borrow: {
        assets: assets.filter(asset => asset.userBorrowBalanceTokens.isGreaterThan(0)),
        marketType: 'borrow',
        breakpoint: 'md',
        columns: ['asset', 'borrowApy', 'userBorrowBalance', 'userPercentOfLimit'],
        initialOrder: {
          orderBy: 'userBorrowBalance',
          orderDirection: 'desc',
        },
      },
    }),
    [assets],
  );

  return (
    <div data-testid={TEST_IDS.tables}>
      {/* Desktop view */}
      <div css={[styles.desktopContainer, hideXlDownCss]}>
        <MarketTable
          {...marketTableProps.supply}
          title={t('account.marketBreakdown.tables.supplyTableTitle')}
        />

        <MarketTable
          {...marketTableProps.borrow}
          title={t('account.marketBreakdown.tables.borrowTableTitle')}
        />
      </div>

      {/* Tablet/Mobile view */}
      <Paper css={[styles.tabletContainer, showXlDownCss]}>
        <div css={styles.tabletHeader}>
          <Typography variant="h4" css={[styles.tabletHeaderTitle, hideMdDownCss]}>
            {t('account.marketBreakdown.tables.tabletTitle')}
          </Typography>

          <ButtonGroup
            css={styles.tabletHeaderButtonGroup}
            buttonLabels={[
              t('account.marketBreakdown.tables.tabletsupplyTabTitle'),
              t('account.marketBreakdown.tables.tabletborrowTabTitle'),
            ]}
            activeButtonIndex={activeTabIndex}
            onButtonClick={setActiveTabIndex}
          />
        </div>

        {activeTabIndex === 0 ? (
          <MarketTable
            key="supply-market-table"
            {...marketTableProps.supply}
            css={styles.tabletMarketTable}
          />
        ) : (
          <MarketTable
            key="borrow-market-table"
            {...marketTableProps.borrow}
            css={styles.tabletMarketTable}
          />
        )}
      </Paper>
    </div>
  );
};

export default Tables;
