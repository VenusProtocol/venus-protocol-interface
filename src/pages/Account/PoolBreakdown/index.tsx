/** @jsxImportSource @emotion/react */
import { Typography } from '@mui/material';
import React from 'react';
import { Pool } from 'types';
import { isFeatureEnabled } from 'utilities';

import Summary from '../Summary';
import Tables from './Tables';
import { useStyles } from './styles';
import TEST_IDS from './testIds';

export interface PoolBreakdownProps {
  pool: Pool;
  className?: string;
}

export const PoolBreakdown: React.FC<PoolBreakdownProps> = ({ pool, className }) => {
  const styles = useStyles();

  return (
    <div className={className}>
      {isFeatureEnabled('isolatedPools') && (
        <div css={styles.title} data-testid={TEST_IDS.title}>
          <Typography css={styles.marketName} variant="h3">
            {pool.name}
          </Typography>
        </div>
      )}

      <Summary pools={[pool]} displayAccountHealth css={styles.summary} />

      <Tables pool={pool} />
    </div>
  );
};

export default PoolBreakdown;
