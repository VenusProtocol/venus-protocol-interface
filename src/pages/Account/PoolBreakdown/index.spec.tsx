import React from 'react';

import { poolData } from '__mocks__/models/pools';
import renderComponent from 'testUtils/renderComponent';

import PoolBreakdown, { PoolBreakdownProps } from '.';
import TEST_IDS from './testIds';

jest.mock('clients/api');

const baseProps: PoolBreakdownProps = {
  pool: poolData[0],
  includeXvs: true,
};

describe('pages/Account/PoolBreakdown', () => {
  it('renders without crashing', () => {
    renderComponent(<PoolBreakdown {...baseProps} />);
  });

  it.each([true, false])(
    'displays stats and tables correctly when includeXvs is %s',
    includeXvs => {
      const { getByTestId } = renderComponent(
        <PoolBreakdown {...baseProps} includeXvs={includeXvs} />,
      );

      expect(getByTestId(TEST_IDS.stats).textContent).toMatchSnapshot();
      expect(getByTestId(TEST_IDS.tables).textContent).toMatchSnapshot();
    },
  );
});
