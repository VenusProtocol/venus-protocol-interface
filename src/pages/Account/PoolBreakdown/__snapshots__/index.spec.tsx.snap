// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/Account/PoolBreakdown Feature flag enabled: isolatedPools displays content correctly 1`] = `"Net APY0%Daily Earnings$0.03Total Supply$1,238,798.65Total Borrow$123.33Borrow limit used:6.4%Limit:$1,926.73.Your safe limit:$1,541.38"`;

exports[`pages/Account/PoolBreakdown Feature flag enabled: isolatedPools displays content correctly 2`] = `"Supplied assetsAssetAPY / LTVBalancesorted descendingCollateralXVS0.17%50%90.00 XVS$115.08USDC5.24%80%100.00 USDC$99.99Sort byBalance​XVSAPY / LTV0.17%50%Balance90.00 XVS$115.08CollateralUSDCAPY / LTV5.24%80%Balance100.00 USDC$99.99CollateralBorrowed assetsAssetAPYBalancesorted descending% of limitBUSD-4.9%50.00 BUSD$50.000%USDT-5.5%40.00 USDT$40.000%Sort byBalance​BUSDAPY-4.9%Balance50.00 BUSD$50.00% of limit0%USDTAPY-5.5%Balance40.00 USDT$40.00% of limit0%AssetsSuppliedBorrowAssetAPY / LTVBalancesorted descendingCollateralXVS0.17%50%90.00 XVS$115.08USDC5.24%80%100.00 USDC$99.99Sort byBalance​XVSAPY / LTV0.17%50%Balance90.00 XVS$115.08CollateralUSDCAPY / LTV5.24%80%Balance100.00 USDC$99.99Collateral"`;

exports[`pages/Account/PoolBreakdown displays content correctly 1`] = `"Supplied assetsAssetAPY / LTVBalancesorted descendingCollateralXVS0.17%50%90.00 XVS$115.08USDC5.24%80%100.00 USDC$99.99Sort byBalance​XVSAPY / LTV0.17%50%Balance90.00 XVS$115.08CollateralUSDCAPY / LTV5.24%80%Balance100.00 USDC$99.99CollateralBorrowed assetsAssetAPYBalancesorted descending% of limitBUSD-4.9%50.00 BUSD$50.000%USDT-5.5%40.00 USDT$40.000%Sort byBalance​BUSDAPY-4.9%Balance50.00 BUSD$50.00% of limit0%USDTAPY-5.5%Balance40.00 USDT$40.00% of limit0%AssetsSuppliedBorrowAssetAPY / LTVBalancesorted descendingCollateralXVS0.17%50%90.00 XVS$115.08USDC5.24%80%100.00 USDC$99.99Sort byBalance​XVSAPY / LTV0.17%50%Balance90.00 XVS$115.08CollateralUSDCAPY / LTV5.24%80%Balance100.00 USDC$99.99Collateral"`;
